lib/grpc/stream.ex

```diff
@@ -190,19 +203,125 @@ defmodule GRPC.Stream do
       raise ArgumentError, "run_with/3 is not supported for unary streams"
     end
 
-    dry_run? = Keyword.get(opts, :dry_run, false)
-
     flow
-    |> Flow.map(fn msg ->
-      if not dry_run? do
-        send_response(from, msg)
-      end
+    |> Flow.map(fn
+      {:ok, msg} ->
+        send_response(from, msg, opts)
+
+        flow
+
+      {:error, %GRPC.RPCError{} = reason} ->
+        send_response(from, reason, opts)
+        flow
 
-      flow
+      {:error, reason} ->
+        msg = GRPC.RPCError.exception(message: "[Error] #{inspect(reason)}")
+        send_response(from, msg, opts)
+        flow
+
+      msg ->
+        send_response(from, msg, opts)
+        flow
     end)
     |> Flow.run()
   end
 
+  @doc """
+  Intercepts and transforms **error tuples** or **unexpected exceptions** that occur
+  within a gRPC stream pipeline.
+
+  `map_error/3` allows graceful handling or recovery from errors produced by previous
+  operators (e.g. `map/2`, `flat_map/2`) or from validation logic applied to incoming data.
+
+  The provided `handler/1` function receives the error reason (or the exception struct)
+  and can either:
+
+    * Return a new error tuple — e.g. `{:error, new_reason}` — to re-emit a modified error.
+    * Return any other value to recover from the failure and continue the pipeline.
+
+  This makes it suitable for both **input validation** and **capturing unexpected runtime errors**
+  in stream transformations.
+
+  ## Parameters
+
+    - `stream` — The input stream or `Flow` pipeline.
+    - `func` — A function that takes an error reason or exception and returns either a new value or an error tuple.
+
+  ## Returns
+
+    - A new stream where all error tuples and raised exceptions are processed by `func/1`.
+
+  ## Examples
+
+      iex> GRPC.Stream.from([1, 2])
+      ...> |> GRPC.Stream.map(fn
+      ...>   2 -> raise "boom"
+      ...>   x -> x
+      ...> end)
+      ...> |> GRPC.Stream.map_error(fn
+      ...>   {:error, _reason} ->
+      ...>     GRPC.RPCError.exception(message: "Validation or runtime error")
+      ...>   msg -> msg
```

Comment:
The docs seem to indicate that the function would only receive errors. Is that the case? If so, there shouldn't be a second clause here.